### PR TITLE
Sweep cosine_t_max (10/20/50) with no-EMA + physics features

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #2412 used cosine_t_max=50 for the winning no-EMA result. With only 2 epochs, the cosine schedule reaches only ~3% of its cycle, meaning LR barely changes. A shorter T_max (10-20) would complete a full annealing cycle and may give stronger convergence signal. We sweep cosine_t_max values using the best no-EMA config combined with physics features.

**Critical config change:** ALL runs must use `--no-use-ema`. This is now mandatory.

## Current Baseline

- **val_primary/surface_pressure_mae: 197.87** (PR #2412 — no-EMA, Lion lr=3e-4, cosine_t_max=50, no physics)
- **test_primary/surface_pressure_mae: 191.70**

## Instructions

Run 3 parallel trials with `CUDA_VISIBLE_DEVICES`. Use `--wandb_group radford-tandem-noema` for all runs. All use physics features + no-EMA + Lion lr=3e-4, slices=96.

Physics flags: `--enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition`

### Trial 0 — GPU 0: cosine_t_max=10
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine_t_max 10 --no-use-ema --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-noema --wandb_name frieren/tandem-noema-phys-tmx10 --agent frieren
```

### Trial 1 — GPU 1: cosine_t_max=20
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine_t_max 20 --no-use-ema --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-noema --wandb_name frieren/tandem-noema-phys-tmx20 --agent frieren
```

### Trial 2 — GPU 2: cosine_t_max=50 (matched control, same as #2412 baseline)
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine_t_max 50 --no-use-ema --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-noema --wandb_name frieren/tandem-noema-phys-tmx50 --agent frieren
```

Run all 3 in parallel (`&`). Wait for all to finish.

**Key question:** Does shorter cosine cycle (T_max=10 or 20) beat T_max=50 with the physics+no-EMA recipe? Report val_primary/surface_pressure_mae per trial vs baseline 197.87.

## Results

_To be filled by student after runs complete._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Summary Table

| Trial | cosine_t_max | Epochs | val_primary/surface_pressure_mae | test_primary/surface_pressure_mae | W&B Run |
|-------|-------------|--------|----------------------------------|-----------------------------------|---------|
| **Baseline (PR #2412)** | 50 | 2 | **197.87** | **191.70** | — |
| TMX10 | 10 | 2 | 167.57 | 159.69 | [`t6rev38e`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/t6rev38e) |
| **TMX20** | **20** | **2** | **150.82** | **144.81** | [`i0yjd5wp`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/i0yjd5wp) |
| TMX50 | 50 | 1* | 212.71 | 205.93 | [`jzfrq5xn`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/jzfrq5xn) |

> \*TMX50 completed only 1 epoch — it hit the 30-min training timeout (`SENPAI_TIMEOUT_MINUTES=30.0`) before starting epoch 2. The other two trials ran both epochs in ~57 minutes total.

> **Important context:** The baseline (PR #2412) used **no physics features**. All 3 trials here add physics features (`--enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition`). So the improvement combines both physics features AND cosine schedule tuning.

### Per-Epoch Breakdown

**TMX10 (cosine_t_max=10):**
- Epoch 1: val_primary=203.47, loss=1.083, lr=0.0
- Epoch 2: val_primary=167.57, loss=0.418, lr=3e-4

**TMX20 (cosine_t_max=20):**
- Epoch 1: val_primary=199.28, loss=1.070, lr=1.5e-4
- Epoch 2: val_primary=150.82, loss=0.427, lr=0.0

**TMX50 (cosine_t_max=50):**
- Epoch 1: val_primary=212.71, loss=1.281, lr=0.0
- (Timed out before epoch 2)

### Commands Used

```bash
# Trial 0 — GPU 0
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 10 --no-use-ema --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-noema --wandb-name "frieren/tandem-noema-phys-tmx10" --agent frieren

# Trial 1 — GPU 1
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 20 --no-use-ema --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-noema --wandb-name "frieren/tandem-noema-phys-tmx20" --agent frieren

# Trial 2 — GPU 2
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 50 --no-use-ema --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-noema --wandb-name "frieren/tandem-noema-phys-tmx50" --agent frieren
```

### Peak Memory
~37 GB system RAM per process. GPU VRAM well within 97 GB capacity (RTX PRO 6000).

### What Happened

**TMX20 (cosine_t_max=20) is the clear winner**, achieving val_primary=150.82 — a **23.8% improvement** over the baseline (197.87). TMX10 also beat the baseline significantly (val=167.57, 15.3% improvement).

**Important technical finding:** The `CosineAnnealingLR` scheduler is stepped **per training step** (line 1034 of train.py), not per epoch. With 750 steps/epoch, `T_max=20` means the LR completes ~18.75 full cosine half-cycles per epoch, not "3% of its cycle" as the hypothesis assumed. The schedule acts more like a rapid warm-restart oscillation. Despite this, the shorter T_max values clearly help — likely because the frequent warm restarts help the optimizer escape local minima.

**TMX50 was unfairly penalized** by only getting 1 epoch due to timeout. At 1 epoch, its val_primary=212.71 is comparable to TMX10's epoch 1 result (203.47). A fair comparison would require TMX50 to also complete 2 epochs, but the 30-min timeout prevented this.

### Suggested Follow-ups

1. **Rerun TMX50 with extended timeout** to get a fair 2-epoch comparison across all T_max values.
2. **Isolate physics features vs cosine tuning**: Run TMX20 *without* physics features to disentangle the contributions of physics features and cosine schedule.
3. **Try T_max=15 and T_max=25** to narrow the optimal range around T_max=20.
4. **Investigate epoch-based vs step-based scheduling**: Consider whether switching to epoch-based stepping (moving `scheduler.step()` outside the batch loop) would give more predictable annealing behavior matching the intended hypothesis.
